### PR TITLE
Vars block and row_vars

### LINKED ISSUE #166: `vars` block
## Background

Currently, the `query` attribute for content blocks allows users to query / transform the data from the context and store it under `.query_result` for convenience. Using `query` is not required but simplifies data requests in Go templates. 

This approach has 2 limitations:
- `query` can be defined only in the `content` block, which limits its scope. The query must be copied-pasted if the same data transformation is required in multiple content blocks.
- users can specify only one query and the result of the query is stored under `.query_result`, making it impossible to have "variables" that simplify data access

### The high-level requirements

- an ability to define data transformations (as JQ queries, as some other code later) in the template blocks.
  - the definition should be explicit about the language of the transformation logic.
- an ability to define multiple data transformations at the same time
- an ability to reuse data transformation results: the result of the transformation defined by the parent should be available for the children to use (similar to a variable scope).
- ability to easily piggyback on the syntax for defining the content-block-specific transformations (for example, for [#169](https://github.com/blackstork-io/fabric/issues/169)).

### Explicit marking of the data transformation logic

In FCL, instead of using `query` attribute with a plain string, the query string should be wrapped in a function call:

```hcl
x = query_<lang>("<query-string>")
```

Here, `query_<lang>` function only adds some metadata to the query string, to does not "execute" the query during FCL parsing. For example, `query_jq(".data.elasticsearch.alerts")` defines a JQ query.

### New `vars` block

We introduce a new `vars` block:

  ```hcl
  vars {
    a = "some"
    b = query_jq(".data.other")
  }
  ```

The block defines variables, set in the context under `.vars.a` and `.vars.b` keys. The values of the variables are either static (`"some"` or any structured data) or the results of the data mutation (the results of executing `.data.other` JQ query against the context).

### Renaming `query` into `local_var`

We rename `query` attribute into `local_var` attribute.

Similar to a variable definition, `local_var` can accept any data or a data transformation call. Similar to `query`, `local_var` attribute can be defined only inside the content blocks.

`local_var` attribute is a syntactic sugar that can be translated into

```hcl
vars {
  local = <value>
}
```

with the value stored under `.vars.local` key in the context.

### Which blocks support `vars` block 

The `vars` block can be defined in the `document`, `section`, and `content` blocks. If defined in the parent blocks, the variables are available to all children as part of the context.

For example:

```hcl
section {
  vars {
    foo = query_jq(".data.inline.hosts[0]"),

    bar = query_jq(
      <<-EOT
        .data.elasticsearch.alerts.hits.hits[0]._id
      EOT
    )
  }
  content text {
    value = "The first host is {{ .vars.foo.name }}"
  }
  content text {
    value = "The first alert id is {{ .vars.bar }}"
  }
}
```

### Using `vars` in `ref` blocks

If `vars` block is defined inside `<block-type> ref` block, the variables defined inside the block _extend_ the values defined in the referenced block. 

For example,
```hcl
content text "foo" {
  vars {
    a = 1
  }
  value = "{{ .vars.a }}"
}

content ref "bar" {
  base = content.text.foo

  vars {
    b = 2
  }
  value = "{{ .vars.a }} and {{ .vars.b }}"

}
```
will render "1 and 2". Note, that the variables with the same name are overwritten.

### `data.inline` data source becomes redundant

Since `vars` block supersedes the `data.inline` data source, it can be removed.

## References
- https://gist.github.com/traut/029ce180869364891d176307de8ea06e

### PR DESCRIPTION
**Pull request by Andrew-Morozko**
Resolves #166 #191 #169

- [x] `query_jq` function, available only within `vars`
   Works even if called in arbitrarily nested FCL lists/maps
- [x] ordered query evaluation, references to other queries
    Queries have access to variables in the `vars` block in the order of their definition, access to other vars from parent scopes and `ref base` blocks. If a variable has the same name as in base - it takes its place in evaluation order (akin to hoisting)
- [x] `vars` in documents
- [x] `vars` in sections and content
- [x] fix error reporting location for gojq parse errors (now diagnostic subject points to the query itself, not its attribute)
- [x] load `.env` files
- [x] expose env vars as fcl variables: `myshell = "I use ${env.SHELL} shell"`
- [x] interaction with `ref` blocks
- [x] local var
- [x] remove `from_env_variable` func
-----
Leftover tasks (to be dealt with in another PR)
- [ ] `query` and `.query_result` removal (right now they are reimplemented with the new query engine and are executed after `vars` evaluation)
- [ ] `data.inline` removal
-----
### New tools
- [x]  `pkg/encapsulator`: a helper for casting back and forth between native go types and `cty.CapsuleValue`s in a typesafe manner
- [x] `plugin.DataPath` for setting/getting values from `plugin.Data` structs via a path (index operations on nested maps/lists with human-readable errors)
- [x] `utils.EvalContextByVar` and `utils.EvalContextByFunc` - get the eval context that defines a certain variable or function
- [x] `uitls.MapMap` and `uitls.MapMapErr`: helpers for mapping a function over map values.
- [x] `utils.OnceVal`: same as `sync.OnceVal`, except natively understands diagnostic semantics. If the first run returns an error - subsequent runs return `RepeatedError`

-----

# Refactoring to support row_vars
*Decided to do it before merging in vars, because some changes to the vars evaluation are needed*

Now there is a `definitions.DataCtxEvalNeeded` interface. All values that need dataCtx to be evaluated (such as `jq_query` and `rows_var`) implement it. This deferred evaluation occurs when the cty.Value is converted into another format:
* for `jq_query` in var blocks this happens when the variable is evaluated and converted to `plugin.Data`
* for `rows_var` this will happen when the value of `rows_var` is converted for grpc transmission.

In order to prevent duplication of the tricky evaluation/encapsulation logic there now exists `pkg/ctyencoder`: a generic transformer from cty to user-chosen type. For us, it's `plugin.Data` and `pluginapiv1.CtyValue` (the grpc type).

`pkg/ctyencoder` provides better error reporting: errors in nested cty objects now come with paths (sequence of indexing operations).
To implement this `pkg/diagnostics` has been refactored to allow easily adding `Extra` values to diagnostics. `diagnostics.PrintDiags` uses this extra information to improve diagnostics before they are printed.
Circular ref detection diagnostics no longer work by writing the traceback as the function returns, now they modify the corresponding Extra, and the printer is the one responsible for outputting the traceback.

Support for passing around `plugin.Data` inside of cty type system and even outside of it (via grpc or with plugin.Data encoding)
Simplified `cty` and `data` protobuf types, now there are fewer nested one-field structs, fewer messages overall, and less code.

TODO: 
- [x] actually evaluate the encapulated `plugin.Data` in content provider args. `eval/dataquery/deferred_plugin_data.go` is the WIP type for `rows_var`
- [x] Replace `encodeCtyValue` with an evaluator capable of providing dataCtx and ctx for deferred evaluation code `plugin/pluginapi/v1/cty_value_encoder.go`
- [x] decide where each component should live

### CODE REVIEW

**Comment by Andrew-Morozko** [comment]
One question regarding `ref`s:

```hcl
document "hello" {
  content text "base" {
    vars {
      a = "original"
      b = query_jq(".vars.a")
    }
    value = "{{toPrettyJson .vars}}"
  }
  content ref "ref" {
    base = document.hello.content.text.base
    vars {
      a = "redefined"
    }
  }
}
```

I assume that here the base block outputs `"a": "original", "b": "original"` and the ref block outputs `"a": "redefined", "b": "original"` (and not `"a": "redefined", "b": "redefined"`). Otherwise reasoning about values of `vars` becomes too difficult and non-local.

In other words, shadowing happens after evaluation, not before. Am I correct? @traut

**Comment by traut** [comment]
@Andrew-Morozko made the issue for env vars -- https://github.com/blackstork-io/fabric/issues/191

**Comment by traut** [comment]
> I assume that here the base block outputs "a": "original", "b": "original" and the ref block outputs "a": "redefined", "b": "original" (and not "a": "redefined", "b": "redefined"). Otherwise reasoning about values of vars becomes too difficult and non-local.
> In other words, shadowing happens after evaluation, not before.

wouldn't it be easier to shadow before evaluation, though? We collapse the var definitions first, and evaluate after. In your example, I would kind of expect `"a": "redefined", "b": "redefined"`. It would also follow the theme that the referenced block is not _evaluated_ but the referencing one is. 

I might be off here, but I can't think of when this breaks the expected output. What do you think?

**Comment by traut** [comment]
damn, those edge cases 😅 

> This should output "a": "redefined", "b": "original", I guess? So, the redefinitions happen in order of variable declaration.

I think you're right. 

> must output {"a": {"b": "val", "c": null}}, since the a is not set when query_jq is evaluated

that's totally reasonable!

**Comment by Andrew-Morozko** [comment]
> wouldn't it be easier to shadow before evaluation, though?

It is kinda easier, but in the end it's just a question of what is more expected by the user (and easier to describe in the docs 😉). I also think that `"a": "redefined", "b": "redefined"` better matches the established `ref` block mechanics, but I also imagined debugging this example:

```hcl
document "hello" {
  content text "base" {
    vars {
      a = ["original"]
      b = query_jq(".vars.a[0]")
    }
    value = "{{toPrettyJson .vars}}"
  }
  content ref "ref" {
    base = document.hello.content.text.base
    vars {
      a = {
        b = "redefined"
      }
    }
  }
}
```

Which would produce this error:
```
Error: Failed to run the query

  on zzz.fabric line 5, in document "hello":
   5:       b = query_jq(".vars.a[0]")

expected an array but got: object ({"b":"redefined"})
```

And the user would look at lines 4 and 5 and not understand where that object came from. 

But perhaps the solution is including a note "this error was triggerd through ref at ...", not special-casing var handling for refs.

**Comment by traut** [comment]
@Andrew-Morozko let's go with `"a": "redefined", "b": "redefined"`, so collapse/shadow the vars and evaluate after. 

You are right, making it clear is very important. I think it's straightforward to explain this way. 

> But perhaps the solution is including a note "this error was triggerd through ref at ...", not special-casing var handling for refs.

Yeah, if the error is in the ref block, we can use different wording or provide more context!

**Comment by traut** [comment]
> Right now ref vars are executed after the base vars, except in cases where var in ref shadows the one in base.

Why split it into two execution steps? To avoid re-executing vars in the base block if there are multiple `ref` blocks?

> ```hcl
> content ref "ref-if-after (this is the current behavior)" {
>   base = document.hello.content.text.base
>   vars {
>     c = "unique to ref"
>     q_b = query_jq(".vars.b") // works as expected
>     a = query_jq(".vars.c") // doesn't have access to c, because "a" overrides the variable in the "base", and therefore is executed before c is set
>   }
> }
> ```

That is caused by the split into two steps. Merging the vars and executing after would resolve this issue: `a` would de redefined and would return `unique to ref`, if I understand correctly

>   // If we evaluate vars of the ref before vars of the base another issue occurs:

Yeah, same problem. It feels like the vars need to be merged before any evaluation happens. Otherwise, we have this implicit order of evaluation between blocks that is a bit confusing.

**Comment by Andrew-Morozko** [comment]
> Is that the one you are leaning toward, too?

Yep, that's the most logical one

File: eval/dataquery/jq.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	val = JqQueryType.ToCty(&JqQuery{
+		query:    query,
+		code:     code,
+		srcRange: expr.Range().Ptr(),
+	})
+	return
+}
+
+var JqQueryType *encapsulator.Codec[JqQuery]
```
Comment (traut) [inline eval/dataquery/jq.go:72]: 🔥

**Review by traut — APPROVED** [review]

**Comment by traut** [comment]
@Andrew-Morozko if `rows_var` is implemented outside the `table` content provider, it makes sense to rename it into more generic `collection_var` or something, so that we can reuse it for other providers

File: eval/vars.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+// Evaluates `variables` and stores the results in `dataCtx` under the key "vars".
+func ApplyVars(ctx context.Context, variables *definitions.ParsedVars, dataCtx plugin.MapData) (diags diagnostics.Diag) {
+	if variables.Empty() {
+		return
+	}
+	var vars plugin.MapData
+
+	varsData := dataCtx["vars"]
```
Comment (dobarx) [inline eval/vars.go:20]: Maybe we should use const for `"vars"`?

**Review by dobarx — APPROVED** [review]
Looks awesome! 🔥

**Comment by Andrew-Morozko** [comment]
Finally implemented row vars. Here's the implementation, including changes discussed on slack.

```hcl
document "test" {
    vars {
        a = 1
        b = query_jq(".vars.a + 1")
        xxx = "xxx"
    }
    content table {
        rows_var = query_jq("[10, 20, 10*(.vars.b + 1)]")

        columns = [
            { header = "{{ .block.col_index }} Value", value = "{{ .block.row }}"},
            { header = "{{ .block.col_index }} Index", value = "{{ .block.row_index }}"},
            { header = "{{ .block.col_index }} ValueFromContext {{ .vars.xxx }}", value = "{{ .vars.xxx }}"},
            { header = "{{ .block.col_index }} StaticValue", value = "foo"},
        ]
    }
}
```
Result:
```md
|1 Value|2 Index|3 ValueFromContext xxx|4 StaticValue|
|---|---|---|---|
|10|1|xxx|foo|
|20|2|xxx|foo|
|30|3|xxx|foo|
```

File: eval/vars.go (hunk 2 above)
Comment (Andrew-Morozko) [reply eval/vars.go:20]: Unfortunantly all maps [are actually pointers](https://groups.google.com/g/golang-nuts/c/SjuhSYDITm4/m/jnrp7rRxDQAJ), even more of a pointer type than a slice, and so [you can't even create a const map](https://go.dev/play/p/ZPv7glrXOT8)

**Pull request merged**